fix: content scrolling so it stops at bottom of viewable area instead…

s9tpepper opened the pull request:
… of at the top.

Before fix the content would scroll on the last page to the top of the viewable area, leaving a large blank space on the last page. This fixes the scrolling so the bottom of the last page stops at the bottom of the viewable area. Adjustment is made to the scrollbar thumb so it stops at the bottom of the scroll track.

Commit:
fix: content scrolling so it stops at bottom of viewable area instead of at the top

joshka commented on src/scroll_view.rs:
```diff
@@ -113,9 +113,21 @@ impl StatefulWidget for ScrollView {
+        // also, ensure that the scrolling stops with the end of the content at the
+        // bottom of the visible area
+        let max_y_offset = self.size.height - area.height;
+        let next_y_offset = y.min(self.buf.area.height.saturating_sub(1));
+        let y_offset = if next_y_offset > max_y_offset {
+            max_y_offset
+        } else {
+            next_y_offset
+        };
+
         x = x.min(self.buf.area.width.saturating_sub(1));
-        y = y.min(self.buf.area.height.saturating_sub(1));
+        y = y.min(y_offset);
```
I suspect this can be simplified to:
```rust
let min_height = area.height.min(1); // show at least one row
let max_offset = self.size.height.saturating_sub(min_height);
y = y.min(max_offset);
```

joshka reviewed (changes requested):
Can you please add unit tests (similar to the ones below) and doc comments that talk about how this works (and which will show up in the docs.rs page).

If you write the tests first, this should be able to confirm that the simplification is the same as what you had (I'm not 100% sure it is).

joshka commented:
https://github.com/joshka/tui-scrollview/pull/30 changes the implementation a bunch, so I'm going to redo this and credit you for the fix.

s9tpepper commented:
Ok sounds good! I was trying to add a test to try the simplification but I couldn’t figure out how to get the scroll thumb to size correctly, it kept drawing two cells tall instead of one.

s9tpepper committed:
Merge branch 'joshka:main' into main

Commit:
test: added scroll_to_bottom test to verify not scrolling past the last line of content

Commit:
chore: fixed clippy hints on unneeded returns

s9tpepper commented:
I merged the latest changes to my fork and I added a `scroll_to_bottom` test to verify the changes that I am proposing.

I also added a fixture that prints the row numbers to make it easier to verify that the scroll position is at the right location.

I simplified the logic around the y position that I updated and I cleaned up some messages that Clippy was complaining about.

Let me know what you think.

joshka commented:
Hey there, it might be some time before I respond to this, but I will get back to it as soon as I can.

joshka committed:
add more tests for scroll position shenanigans

joshka committed:
Merge remote-tracking branch 'origin/main' into s9tpepper/main

joshka commented on src/scroll_view.rs:
```diff
@@ -117,9 +119,16 @@ impl StatefulWidget for ScrollView {
+
         x = x.min(self.buf.area.width.saturating_sub(1));
-        y = y.min(self.buf.area.height.saturating_sub(1));
+        y = y.min(min(next_y_offset, max_y_offset));
```
This is a bit convoluted y.min(min(a calculation with min from above))

joshka commented on src/scroll_view.rs:
```diff
@@ -117,9 +119,16 @@ impl StatefulWidget for ScrollView {
         // ensure that we don't scroll past the end of the buffer in either direction
+        // also, ensure that the scrolling stops with the end of the content at the
+        // bottom of the visible area
+        let max_y_offset = self.buf.area.height.saturating_sub(area.height);
```
I changed this to a saturating sub as one of the tests was panicking due to underflow here.

joshka reviewed (changes requested):
Hey, I just took a look - thanks for reminding me to come back to this.

I merged main into your branch and added some more tests. This now uses the following more convenient test data (which makes it easier to see when there's problems with both rows and columns):

https://github.com/joshka/tui-scrollview/blob/533150d021055bd760b4e14f46e6978662e1e7cb/src/scroll_view.rs#L229-L238

It works fine when you don't have a horizontal scrollbar:

https://github.com/joshka/tui-scrollview/blob/533150d021055bd760b4e14f46e6978662e1e7cb/src/scroll_view.rs#L438-L442

But fails when you do have a horizontal scrollbar (the last row is never shown):

https://github.com/joshka/tui-scrollview/blob/533150d021055bd760b4e14f46e6978662e1e7cb/src/scroll_view.rs#L395-L400

This also should handle scrolling to the right as well.

https://github.com/joshka/tui-scrollview/blob/533150d021055bd760b4e14f46e6978662e1e7cb/src/scroll_view.rs#L341-L346

Note the failing tests likely have the wrong scrollbar positions, but the visible data should likely be correct. Hopefully the tests should guide you towards the right solution here.
